modules/apigee_edge_apidocs/tests/src/Kernel/ApidocEntityRevisionsAccessTest.php

```diff
@@ -0,0 +1,200 @@
+
+    $this->entityTypeManager = $this->container->get('entity_type.manager');
+    $this->entityTypeStorage = $this->entityTypeManager->getStorage('apidoc');
+
+    // Create a published apidoc.
+    $apidoc = ApiDoc::create([
+      'name' => 'API 1',
+      'description' => 'Test API v1',
+      'spec' => NULL,
+      'api_product' => NULL,
+      'status' => 1,
+    ]);
+    $apidoc->save();
+    $this->apidocV1Id = $apidoc->getRevisionId();
+
+    // Create a new revision.
+    $apidoc->setDescription('Test API v2');
+    $apidoc->setRevisionLogMessage('v2');
+    $apidoc->setNewRevision();
+    $apidoc->save();
+    $this->apidocV2Id = $apidoc->getRevisionId();
+
+    $this->apidoc = $apidoc;
+
+    // Discard user 1, we will not need it because it bypasses access control.
+    $this->createUser();
+  }
+
+  /**
+   * Test ApiDocs revision access as anonymous.
+   */
+  public function testApiDocRevisionsAccessAnon() {
+    $entity_v1 = $this->entityTypeStorage->loadRevision($this->apidocV1Id);
+
+    $tests = [
+      'view' => 'Anonymous should not be able to view an unpublished revision.',
+      'update' => 'Anonymous should not be able to update a revision.',
+    ];
+
+    foreach ($tests as $op => $message) {
+      $this->assertFalse($entity_v1->access($op), $message);
+    }
+  }
+
+  /**
+   * Test ApiDocs revision access a logged in user.
+   */
+  public function testApiDocRevisionsAccessLoggedIn() {
+    $user = $this->createUser([]);
+    \Drupal::currentUser()->setAccount($user);
+
+    $entity_v1 = $this->entityTypeStorage->loadRevision($this->apidocV1Id);
+
+    $tests = [
+      'view' => 'LoggedIn should not be able to view an unpublished revision.',
+      'update' => 'LoggedIn should not be able to update a revision.',
+    ];
+
+    foreach ($tests as $op => $message) {
+      $this->assertFalse($entity_v1->access($op, $user), $message);
+    }
+  }
+
+  /**
+   * Test ApiDocs revision access as a logged in user with some permissions.
+   */
+  public function testApiDocRevisionsAccessPermissions() {
+    $user = $this->createUser([
+      'view published apidoc entities',
+      'view unpublished apidoc entities',
+      'view all apidoc revisions',
+      'edit apidoc entities',
+      'revert all apidoc revisions',
+    ]);
+    \Drupal::currentUser()->setAccount($user);
+
+    $entity_v1 = $this->entityTypeStorage->loadRevision($this->apidocV1Id);
+
+    $tests = [
+      'view' => 'User should be able to view an unpublished revision.',
+      'update' => 'User should be able to update a revision.',
+    ];
+
+    foreach ($tests as $op => $message) {
+      $this->assertTrue($entity_v1->access($op, $user), $message);
+    }
+  }
+
+  /**
+   * Test ApiDocs revision access as a logged in user with admin permissions.
+   */
+  public function testApiDocRevisionsAccessAdmin() {
+    $user = $this->createUser([
+      'administer apidoc entities',
+    ]);
+    \Drupal::currentUser()->setAccount($user);
```

Comment:
Inject and use the current_user service or use the account_switcher current. I do not think that this is the right way to switch user in tests.

Reply:
Cool service, I hadn't run into it before. Code has been updated to use `$this->container->get('account_switcher')->switchTo($user);`